libtiny_client/src/state.rs

```diff
@@ -735,35 +735,57 @@ fn parse_servername(params: &[String]) -> Option<String> {
     }
 }
 
+/// Parse the server name from prefix
+fn parse_server_pfx(pfx: Option<&Pfx>) -> Option<String> {
+    if let Some(Pfx::Server(server_name)) = pfx {
+        Some(server_name.to_owned())
+    } else {
+        None
+    }
+}
+
+/// Parse server name from RPL_YOURHOST reply
+/// or fallback to using the server name inside Pfx::Server
+fn parse_servername(msg: &Msg) -> Option<String> {
+    use wire::Cmd::Reply;
+    if let Reply { params, .. } = &msg.cmd {
+        parse_yourhost_msg(&params).or(parse_server_pfx(msg.pfx.as_ref()))
+    } else {
+        parse_server_pfx(msg.pfx.as_ref())
+    }
```

Comment:
In the call site we already know that this is a reply, with code 002, so this branch will never be taken. We can remove it completely if you pass `params` and `pfx` to `parse_servername`, instead of `msg`. Something like:

```
fn parse_servername(pfx, params) -> Option<String> {
    parse_yourhost_msg(&params).or_else(|| parse_server_pfx(pfx))
}
```

Note that I used `or_else` instead of `or`, if you use `or` you'll evaluate the argument even when `parse_yourhost_msg(...)` will return `Some(...)`. You can evaluate it lazily by using `or_else` and a closure.